### ISSUE TITLE
Crash - Running out of memory on fly.io when downloading the budget

### ISSUE THREAD

**Issue by SylvainMartel** [actualbudget/actual-server#54]
I am trying to download my budget on a new computer but the app keeps crashing on fly.io.  The only way I managed to make it work was to "sacle" the vm to 512mb for the download, then I was able to downscale it to 256 again.  Any ideas why the download seems to require more memory?(which fall out of the free tier if I'm correct.)

The app will crash-loop forever with 256Mb, here is the log from fly.io

```
 2022-06-11T13:17:24.258 runner[7c4ae106] yyz [info] Starting virtual machine
2022-06-11T13:17:24.453 app[7c4ae106] yyz [info] Starting init (commit: e3eb6d2)...
2022-06-11T13:17:24.467 app[7c4ae106] yyz [info] Mounting /dev/vdc at /data w/ uid: 0, gid: 0 and chmod 0755
2022-06-11T13:17:24.478 app[7c4ae106] yyz [info] Preparing to run: `/usr/bin/tini -g -- node app.js` as root
2022-06-11T13:17:24.504 app[7c4ae106] yyz [info] 2022/06/11 13:17:24 listening on [fdaa:0:5f82:a7b:8a0f:0:debd:2]:22 (DNS: [fdaa::3]:53)
2022-06-11T13:17:24.506 app[7c4ae106] yyz [info] [WARN tini (523)] Tini is not running as PID 1 and isn't registered as a child subreaper.
2022-06-11T13:17:24.506 app[7c4ae106] yyz [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
2022-06-11T13:17:24.506 app[7c4ae106] yyz [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
2022-06-11T13:17:25.224 app[7c4ae106] yyz [info] Initializing Actual with user file dir: /data/user-files
2022-06-11T13:17:25.226 app[7c4ae106] yyz [info] Listening on 0.0.0.0:5006...
2022-06-11T13:19:28.352 app[7c4ae106] yyz [info] [ 123.971282] Out of memory: Killed process 528 (node) total-vm:947256kB, anon-rss:200868kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:1420kB oom_score_adj:0
2022-06-11T13:19:28.637 app[7c4ae106] yyz [info] Main child exited normally with code: 137
2022-06-11T13:19:28.638 app[7c4ae106] yyz [info] Starting clean up.
2022-06-11T13:19:28.654 app[7c4ae106] yyz [info] Umounting /dev/vdc from /data
2022-06-11T13:19:35.681 runner[7c4ae106] yyz [info] Starting instance
2022-06-11T13:19:35.828 runner[7c4ae106] yyz [info] Configuring virtual machine
2022-06-11T13:19:35.833 runner[7c4ae106] yyz [info] Pulling container image
2022-06-11T13:19:36.032 runner[7c4ae106] yyz [info] Unpacking image
2022-06-11T13:19:36.044 runner[7c4ae106] yyz [info] Preparing kernel init
2022-06-11T13:19:36.239 runner[7c4ae106] yyz [info] Setting up volume 'actual_data'
2022-06-11T13:19:36.244 runner[7c4ae106] yyz [info] Opening encrypted volume
2022-06-11T13:19:37.599 runner[7c4ae106] yyz [info] Configuring firecracker
2022-06-11T13:19:37.638 runner[7c4ae106] yyz [info] Starting virtual machine
2022-06-11T13:19:37.807 app[7c4ae106] yyz [info] Starting init (commit: e3eb6d2)...
2022-06-11T13:19:37.829 app[7c4ae106] yyz [info] Mounting /dev/vdc at /data w/ uid: 0, gid: 0 and chmod 0755
2022-06-11T13:19:37.833 app[7c4ae106] yyz [info] Preparing to run: `/usr/bin/tini -g -- node app.js` as root
2022-06-11T13:19:37.858 app[7c4ae106] yyz [info] [WARN tini (523)] Tini is not running as PID 1 and isn't registered as a child subreaper.
2022-06-11T13:19:37.858 app[7c4ae106] yyz [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
2022-06-11T13:19:37.858 app[7c4ae106] yyz [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
2022-06-11T13:19:37.860 app[7c4ae106] yyz [info] 2022/06/11 13:19:37 listening on [fdaa:0:5f82:a7b:8a0f:0:debd:2]:22 (DNS: [fdaa::3]:53)
2022-06-11T13:19:38.552 app[7c4ae106] yyz [info] Initializing Actual with user file dir: /data/user-files
2022-06-11T13:19:38.553 app[7c4ae106] yyz [info] Listening on 0.0.0.0:5006...
2022-06-11T13:19:45.440 app[7c4ae106] yyz [info] [ 7.702757] Out of memory: Killed process 528 (node) total-vm:913960kB, anon-rss:201324kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:1392kB oom_score_adj:0
2022-06-11T13:19:45.852 app[7c4ae106] yyz [info] Main child exited normally with code: 137
2022-06-11T13:19:45.852 app[7c4ae106] yyz [info] Starting clean up.
2022-06-11T13:19:45.867 app[7c4ae106] yyz [info] Umounting /dev/vdc from /data
2022-06-11T13:19:53.123 runner[7c4ae106] yyz [info] Starting instance
2022-06-11T13:19:53.276 runner[7c4ae106] yyz [info] Configuring virtual machine
2022-06-11T13:19:53.284 runner[7c4ae106] yyz [info] Pulling container image
2022-06-11T13:19:53.486 runner[7c4ae106] yyz [info] Unpacking image
2022-06-11T13:19:53.496 runner[7c4ae106] yyz [info] Preparing kernel init
2022-06-11T13:19:53.686 runner[7c4ae106] yyz [info] Setting up volume 'actual_data'
2022-06-11T13:19:53.690 runner[7c4ae106] yyz [info] Opening encrypted volume
2022-06-11T13:19:54.868 runner[7c4ae106] yyz [info] Configuring firecracker
2022-06-11T13:19:54.922 runner[7c4ae106] yyz [info] Starting virtual machine
2022-06-11T13:19:55.106 app[7c4ae106] yyz [info] Starting init (commit: e3eb6d2)...
2022-06-11T13:19:55.124 app[7c4ae106] yyz [info] Mounting /dev/vdc at /data w/ uid: 0, gid: 0 and chmod 0755
2022-06-11T13:19:55.133 app[7c4ae106] yyz [info] Preparing to run: `/usr/bin/tini -g -- node app.js` as root
2022-06-11T13:19:55.159 app[7c4ae106] yyz [info] 2022/06/11 13:19:55 listening on [fdaa:0:5f82:a7b:8a0f:0:debd:2]:22 (DNS: [fdaa::3]:53)
2022-06-11T13:19:55.160 app[7c4ae106] yyz [info] [WARN tini (523)] Tini is not running as PID 1 and isn't registered as a child subreaper.
2022-06-11T13:19:55.160 app[7c4ae106] yyz [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
2022-06-11T13:19:55.161 app[7c4ae106] yyz [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
2022-06-11T13:19:55.840 app[7c4ae106] yyz [info] Initializing Actual with user file dir: /data/user-files
2022-06-11T13:19:55.841 app[7c4ae106] yyz [info] Listening on 0.0.0.0:5006... 
```

**Comment by kinwolfqc** [comment]
Memory used by the app on fly.io, at idle, goes from 202Mb to 219Mb out of 232Mb available.
The problem is easily reproducible in my case by deleting the local copy and trying to re-download it from the app main screen.

**Comment by shall0pass** [comment]
![2022-07-25 14_56_51-greasyhandsombug · Metrics · Fly - Brave](https://user-images.githubusercontent.com/20625555/180863782-4c956ff3-8c67-43dd-a719-b9a9e817571a.png)

$ fly logs -a greasyhandsombug

Waiting for logs...

2022-07-24T23:50:49.829 app[fa31a9e5] ord [info] Making backup

2022-07-25T00:05:49.860 app[fa31a9e5] ord [info] Making backup

2022-07-25T00:20:49.893 app[fa31a9e5] ord [info] Making backup

2022-07-25T00:35:49.924 app[fa31a9e5] ord [info] Making backup

2022-07-25T00:50:49.955 app[fa31a9e5] ord [info] Making backup

2022-07-25T01:05:49.986 app[fa31a9e5] ord [info] Making backup

2022-07-25T01:20:50.018 app[fa31a9e5] ord [info] Making backup

2022-07-25T01:35:50.051 app[fa31a9e5] ord [info] Making backup

2022-07-25T01:50:50.078 app[fa31a9e5] ord [info] Making backup

2022-07-25T02:05:50.112 app[fa31a9e5] ord [info] Making backup

2022-07-25T02:20:50.143 app[fa31a9e5] ord [info] Making backup

2022-07-25T02:35:50.174 app[fa31a9e5] ord [info] Making backup

2022-07-25T02:50:50.208 app[fa31a9e5] ord [info] Making backup

2022-07-25T03:05:50.237 app[fa31a9e5] ord [info] Making backup

2022-07-25T03:20:50.268 app[fa31a9e5] ord [info] Making backup

2022-07-25T03:35:50.299 app[fa31a9e5] ord [info] Making backup

2022-07-25T03:50:50.326 app[fa31a9e5] ord [info] Making backup

2022-07-25T04:05:50.357 app[fa31a9e5] ord [info] Making backup

2022-07-25T04:20:50.388 app[fa31a9e5] ord [info] Making backup

2022-07-25T04:35:50.417 app[fa31a9e5] ord [info] Making backup

2022-07-25T04:50:50.448 app[fa31a9e5] ord [info] Making backup

2022-07-25T05:05:50.479 app[fa31a9e5] ord [info] Making backup

2022-07-25T05:20:50.509 app[fa31a9e5] ord [info] Making backup

2022-07-25T05:35:50.540 app[fa31a9e5] ord [info] Making backup

2022-07-25T05:50:50.571 app[fa31a9e5] ord [info] Making backup

2022-07-25T06:05:50.601 app[fa31a9e5] ord [info] Making backup

2022-07-25T06:20:50.631 app[fa31a9e5] ord [info] Making backup

2022-07-25T06:35:50.662 app[fa31a9e5] ord [info] Making backup

2022-07-25T06:50:50.692 app[fa31a9e5] ord [info] Making backup

2022-07-25T07:05:50.723 app[fa31a9e5] ord [info] Making backup

2022-07-25T07:20:50.753 app[fa31a9e5] ord [info] Making backup

2022-07-25T07:35:50.783 app[fa31a9e5] ord [info] Making backup

2022-07-25T07:50:50.814 app[fa31a9e5] ord [info] Making backup

2022-07-25T08:05:50.844 app[fa31a9e5] ord [info] Making backup

2022-07-25T08:20:50.874 app[fa31a9e5] ord [info] Making backup

2022-07-25T08:35:50.904 app[fa31a9e5] ord [info] Making backup

2022-07-25T08:50:50.935 app[fa31a9e5] ord [info] Making backup

2022-07-25T09:05:50.965 app[fa31a9e5] ord [info] Making backup

2022-07-25T09:20:50.995 app[fa31a9e5] ord [info] Making backup

2022-07-25T09:35:51.026 app[fa31a9e5] ord [info] Making backup

2022-07-25T09:50:51.057 app[fa31a9e5] ord [info] Making backup

2022-07-25T10:05:51.087 app[fa31a9e5] ord [info] Making backup

2022-07-25T10:20:51.115 app[fa31a9e5] ord [info] Making backup

2022-07-25T10:35:51.145 app[fa31a9e5] ord [info] Making backup

2022-07-25T10:50:51.176 app[fa31a9e5] ord [info] Making backup

2022-07-25T11:05:51.203 app[fa31a9e5] ord [info] Making backup

2022-07-25T11:20:51.233 app[fa31a9e5] ord [info] Making backup

2022-07-25T11:35:51.262 app[fa31a9e5] ord [info] Making backup

2022-07-25T11:50:51.292 app[fa31a9e5] ord [info] Making backup

2022-07-25T12:05:51.323 app[fa31a9e5] ord [info] Making backup

2022-07-25T12:20:51.353 app[fa31a9e5] ord [info] Making backup

2022-07-25T12:29:53.810 app[fa31a9e5] ord [info] [51366.859171] Out of memory: Killed process 528 (node) total-vm:1063588kB, anon-rss:195400kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2816kB oom_score_adj:0

2022-07-25T12:29:54.324 app[fa31a9e5] ord [info] Main child exited normally with code: 137

2022-07-25T12:29:54.324 app[fa31a9e5] ord [info] Starting clean up.

2022-07-25T12:29:54.338 app[fa31a9e5] ord [info] Umounting /dev/vdc from /data

2022-07-25T12:30:01.042 runner[fa31a9e5] ord [info] Starting instance

2022-07-25T12:30:01.410 runner[fa31a9e5] ord [info] Configuring virtual machine

2022-07-25T12:30:01.412 runner[fa31a9e5] ord [info] Pulling container image

2022-07-25T12:30:01.778 runner[fa31a9e5] ord [info] Unpacking image

2022-07-25T12:30:01.784 runner[fa31a9e5] ord [info] Preparing kernel init

2022-07-25T12:30:01.934 runner[fa31a9e5] ord [info] Setting up volume 'actual_data'

2022-07-25T12:30:01.936 runner[fa31a9e5] ord [info] Opening encrypted volume

2022-07-25T12:30:02.101 runner[fa31a9e5] ord [info] Configuring firecracker

2022-07-25T12:30:02.180 runner[fa31a9e5] ord [info] Starting virtual machine

2022-07-25T12:30:02.365 app[fa31a9e5] ord [info] Starting init (commit: c86b3dc)...

2022-07-25T12:30:02.383 app[fa31a9e5] ord [info] Mounting /dev/vdc at /data w/ uid: 0, gid: 0 and chmod 0755

2022-07-25T12:30:02.393 app[fa31a9e5] ord [info] Preparing to run: `/usr/bin/tini -g -- node app.js` as root

2022-07-25T12:30:02.418 app[fa31a9e5] ord [info] 2022/07/25 12:30:02 listening on [fdaa:0:79e1:a7b:9adb:1:52d1:2]:22 (DNS: [fdaa::3]:53)

2022-07-25T12:30:02.420 app[fa31a9e5] ord [info] [WARN tini (523)] Tini is not running as PID 1 and isn't registered as a child subreaper.

2022-07-25T12:30:02.420 app[fa31a9e5] ord [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.

2022-07-25T12:30:02.420 app[fa31a9e5] ord [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.

2022-07-25T12:30:03.215 app[fa31a9e5] ord [info] Initializing Actual with user file dir: /data/user-files

2022-07-25T12:30:03.216 app[fa31a9e5] ord [info] Listening on 0.0.0.0:5006...

2022-07-25T18:57:02.416 app[fa31a9e5] ord [info] Loaded spreadsheet from cache (50930 items)

2022-07-25T19:12:02.270 app[fa31a9e5] ord [info] Making backup

2022-07-25T19:27:02.300 app[fa31a9e5] ord [info] Making backup

2022-07-25T19:42:02.325 app[fa31a9e5] ord [info] Making backup       *****<<<<Deleted local copy and redownloaded here.**

2022-07-25T19:55:49.639 app[fa31a9e5] ord [info] [26746.701188] Out of memory: Killed process 528 (node) total-vm:987288kB, anon-rss:195576kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2844kB oom_score_adj:0

2022-07-25T19:55:49.928 app[fa31a9e5] ord [info] Main child exited normally with code: 137

2022-07-25T19:55:49.928 app[fa31a9e5] ord [info] Starting clean up.

2022-07-25T19:55:49.942 app[fa31a9e5] ord [info] Umounting /dev/vdc from /data

2022-07-25T19:55:55.561 runner[fa31a9e5] ord [info] Starting instance

2022-07-25T19:55:55.870 runner[fa31a9e5] ord [info] Configuring virtual machine

2022-07-25T19:55:55.871 runner[fa31a9e5] ord [info] Pulling container image

2022-07-25T19:55:56.089 runner[fa31a9e5] ord [info] Unpacking image

2022-07-25T19:55:56.095 runner[fa31a9e5] ord [info] Preparing kernel init

2022-07-25T19:55:56.213 runner[fa31a9e5] ord [info] Setting up volume 'actual_data'

2022-07-25T19:55:56.216 runner[fa31a9e5] ord [info] Opening encrypted volume

2022-07-25T19:55:56.390 runner[fa31a9e5] ord [info] Configuring firecracker

2022-07-25T19:55:56.422 runner[fa31a9e5] ord [info] Starting virtual machine

2022-07-25T19:55:56.596 app[fa31a9e5] ord [info] Starting init (commit: c86b3dc)...

2022-07-25T19:55:56.615 app[fa31a9e5] ord [info] Mounting /dev/vdc at /data w/ uid: 0, gid: 0 and chmod 0755

2022-07-25T19:55:56.628 app[fa31a9e5] ord [info] Preparing to run: `/usr/bin/tini -g -- node app.js` as root

2022-07-25T19:55:56.653 app[fa31a9e5] ord [info] 2022/07/25 19:55:56 listening on [fdaa:0:79e1:a7b:9adb:1:52d1:2]:22 (DNS: [fdaa::3]:53)

2022-07-25T19:55:56.655 app[fa31a9e5] ord [info] [WARN tini (523)] Tini is not running as PID 1 and isn't registered as a child subreaper.

2022-07-25T19:55:56.655 app[fa31a9e5] ord [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.

2022-07-25T19:55:56.655 app[fa31a9e5] ord [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.

2022-07-25T19:55:57.365 app[fa31a9e5] ord [info] Initializing Actual with user file dir: /data/user-files

2022-07-25T19:55:57.365 app[fa31a9e5] ord [info] Listening on 0.0.0.0:5006...

2022-07-25T19:56:08.935 app[fa31a9e5] ord [info] Loaded spreadsheet from cache (50930 items)

**Comment by rich-howell** [comment]
This is potentially resolved now, there was a merge of https://github.com/actualbudget/actual-server/pull/72 which fixes the server to only download the required files, there is also work being done around the Sync method https://github.com/actualbudget/actual-server/pull/75 so if #72 doesn't resolve this #75 might.

Additionally, there has been some additional commits to help with Garbage collection and memory management.

You could clone main and try this again :)

**Comment by kinwolfqc** [comment]
deploying master as I type this.  Will see if it helps with the memory issue, thanks!

**Comment by rich-howell** [comment]
Hey @SylvainMartel 

This should be resolved in 20.10.25 if it isn't please do let me know and I will re-open this issue

Cheers
Rich
